File: JavaApp/src/main/java/com/charlezz/javaapp/feature/remote/RemoteAdapter.java
package com.charlezz.javaapp.feature.remote;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;

import com.charlezz.javaapp.BR;
import com.charlezz.javaapp.R;
import com.charlezz.javaapp.di.FragmentScope;
import com.charlezz.javaapp.feature.base.BindingPagedListAdapter;
import com.charlezz.javaapp.feature.base.ViewBindingHolder;

import javax.inject.Inject;

@FragmentScope
@SuppressWarnings("unchecked")
public class RemoteAdapter extends BindingPagedListAdapter{

    private static DiffUtil.ItemCallback<Post> diffCallback= new DiffUtil.ItemCallback<Post>() {
        @Override
        public boolean areItemsTheSame(@NonNull Post oldPost, @NonNull Post newPost) {
            return oldPost.getId() == newPost.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull Post oldPost, @NonNull Post newPost) {
            return oldPost.getBody().equals(newPost.getBody())&&
                    oldPost.describeContents() == newPost.describeContents()&&
                    oldPost.getTitle().equals(newPost.getTitle()) &&
                    oldPost.getUserId() == newPost.getUserId();
        }
    };

    @Inject
    public RemoteAdapter() {
        super(diffCallback);
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public ViewBindingHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new ViewBindingHolder(DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()), R.layout.view_post, viewGroup, false));
    }

    @NonNull
    @Override
    protected int getViewModelId(int viewType) {
        return BR.post;
    }
}

File: app/src/main/java/com/charlezz/finalarchitecture/ui/remote/RemoteAdapter.kt
package com.charlezz.finalarchitecture.ui.remote

import android.support.v7.util.DiffUtil
import com.charlezz.finalarchitecture.R
import com.charlezz.finalarchitecture.data.remote.Post
import com.charlezz.finalarchitecture.databinding.ViewRemoteBinding
import com.charlezz.finalarchitecture.ui.base.BaseAdapter
import com.charlezz.finalarchitecture.ui.base.BaseViewHolder

class RemoteAdapter : BaseAdapter<Post, ViewRemoteBinding, RemoteAdapter.RemoteViewHolder>(diffCallback){
    override fun getLayoutId(): Int = R.layout.view_remote

    inner class RemoteViewHolder(binding: ViewRemoteBinding) : BaseViewHolder<ViewRemoteBinding>(binding)

    companion object {
        val diffCallback = object : DiffUtil.ItemCallback<Post>() {
            override fun areContentsTheSame(oldItem: Post?, newItem: Post?): Boolean {
                return if (oldItem != null && newItem != null){
                    oldItem.id == newItem.id &&
                            oldItem.body == newItem.body &&
                            oldItem.title == newItem.title &&
                            oldItem.userId == newItem.userId
                } else {
                    false
                }
            }

            override fun areItemsTheSame(oldItem: Post?, newItem: Post?): Boolean {
                return oldItem == newItem
            }
        }
    }

}